cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,343 @@
+import Control.Exception
+    ( bracket )
+import qualified Data.ByteString.Char8 as BS
+import Data.ByteString.Lazy ()
+import System.Directory
+    ( canonicalizePath, doesFileExist, getTemporaryDirectory, removeDirectoryRecursive )
+import System.FilePath
+    ( (</>), dropFileName, takeFileName )
+import qualified Text.Parsec as P
+
+
+-- | Get the directory where script builds are cached.
+--
+-- @CABAL_DIR\/script-builds\/@
+getScriptCacheDirectoryRoot :: IO FilePath
+getScriptCacheDirectoryRoot = do
+  cabalDir <- getCabalDir
+  return $ cabalDir </> "script-builds"
+
+-- | Get the hash of a script's absolute path)
+--
+-- Two hashes will be the same as long as the absolute paths
+-- are the same.
+getScriptHash :: FilePath -> IO String
+getScriptHash script = showHashValue . hashValue . fromString <$> canonicalizePath script
+
+-- | Get the directory for caching a script build.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to @CABAL_DIR\/script-builds\/@ to get the cache directory.
+getScriptCacheDirectory :: FilePath -> IO FilePath
+getScriptCacheDirectory script = (</>) <$> getScriptCacheDirectoryRoot <*> getScriptHash script
+
+-- | Get the directory for caching a script build and ensure it exists.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to @CABAL_DIR\/script-builds\/@ to get the cache directory.
+ensureScriptCacheDirectory :: Verbosity -> FilePath -> IO FilePath
+ensureScriptCacheDirectory verbosity script = do
+  cacheDir <- getScriptCacheDirectory script
+  createDirectoryIfMissingVerbose verbosity True cacheDir
+  return cacheDir
+
+-- | What your command should do when no targets are found.
+data AcceptNoTargets
+  = RejectNoTargets -- ^ die on 'TargetSelectorNoTargetsInProject'
+  | AcceptNoTargets -- ^ return a default 'TargetSelector'
+  deriving (Eq, Show)
+
+-- | Information about the context in which we found the 'TargetSelector's.
+data TargetContext
+  = ProjectContext -- ^ The target selectors are part of a project.
+  | GlobalContext  -- ^ The target selectors are from the global context.
+  | ScriptContext FilePath Executable
+  -- ^ The target selectors refer to a script. Contains the path to the script and
+  -- the executable metadata parsed from the script
+  deriving (Eq, Show)
+
+-- | Determine whether the targets represent regular targets or a script
+-- and return the proper context and target selectors.
+-- Die with an error message if selectors are valid as neither regular targets or as a script.
+--
+-- In the case that the context refers to a temporary directory,
+-- delete it after the action finishes.
+withContextAndSelectors
+  :: AcceptNoTargets     -- ^ What your command should do when no targets are found.
+  -> Maybe ComponentKind -- ^ A target filter
+  -> NixStyleFlags a     -- ^ Command line flags
+  -> [String]            -- ^ Target strings or a script and args.
```

Comment:
Have you consider convert this in a sum type like `TargetInfo` (or whatever), tracking in the type system this useful comment? too much work at call sites?

Reply:
`NixStyleFlags a -> [Strings] -> GlobalFlags ->` are what get passed to all command options
`AcceptNoTargets -> Maybe ComponentKind ->` I added as I realized I needed them to account for the differences between commands.

Are you suggesting combing just the first two or all five? I don't think it would be too much work at call sites, but I'm not sure I see a clear benefit either.

Reply:
I was thinking in something like

```haskell

-- With a better name for Targets
data TargetInfo = Targets [String] | ScriptTarget String [String] 

withContextAndSelectors :: AcceptNoTargets -> Maybe ComponentKind -> NixStyleFlags a -> TargetInfo  -> GlobalFlags -> (TargetContext -> ProjectBaseContext -> [TargetSelector] -> IO b) -> IO b
```
Cause [String] is being used for two different things and mixing them should be a type error.
But if `[String]` is used everywhere it is out of the scope of the pr

Reply:
Thinking about this, `withContextAndSelectors` truncates `targetStrings` based on the the `ComponentKind`, which is a bit ad-hoc. Really it's `CmdRun,hs` that knows that only the first one is a target string and should only pass that one.